### PR TITLE
config: generalize to arbitrary resource bootstrapping

### PR DESCRIPTION
**Pull request by sttts**
To be used soon for the root workspace, and org workspaces.

Poor men's CVO.

### CODE REVIEW

File: config/helpers/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		return fmt.Errorf("failed to bootstrap CRDs into root workspace: %w", err)
+	}
+
+	// bootstrap non-crd resources
+	var transformers []TransformFileFunc
+	for _, opt := range opts {
+		transformers = append(transformers, opt.TransformFile)
+	}
+	return wait.PollImmediateInfiniteWithContext(ctx, time.Second, func(ctx context.Context) (bool, error) {
+		// +lint:noerrcheck
```
Comment (marun) [inline config/helpers/bootstrap.go:92]: (No action required) Why is this required here but not above line 77?

**Comment by ncdc** [comment]
>Poor men's CVO.

What is CVO?

**Comment by ncdc** [comment]
Can you rename `dummy-to-make-embed-fs-happy.yaml` to a more inclusive name (e.g. sentinel- or empty- or something like that)? Thanks!

File: config/helpers/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(crdClient.Discovery()))
+
+	if err := wait.PollImmediateInfiniteWithContext(ctx, time.Second, func(ctx context.Context) (bool, error) {
+		if err := configcrds.Create(ctx, crdClient.ApiextensionsV1().CustomResourceDefinitions(), crds...); err != nil {
+			klog.Errorf("failed to bootstrap CRDs: %v", err)
+			return false, nil // keep retrying
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("failed to bootstrap CRDs into root workspace: %w", err)
```
Comment (ncdc) [inline config/helpers/bootstrap.go:83]: Would we ever use this outside of root?

File: config/helpers/bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+// CreateResourceFromFS creates given resource file.
+func CreateResourceFromFS(ctx context.Context, client dynamic.Interface, mapper meta.RESTMapper, filename string, fs embed.FS, transformers ...TransformFileFunc) error {
+	raw, err := fs.ReadFile(filename)
+	if err != nil {
+		return fmt.Errorf("could not read %s: %w", filename, err)
+	}
+
+	if len(raw) == 0 {
+		return nil // ignore dummy files
```
Comment (ncdc) [inline config/helpers/bootstrap.go:128]:
```suggestion
		return nil // ignore empty files
```

File: config/helpers/bootstrap.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		if err := configcrds.Create(ctx, crdClient.ApiextensionsV1().CustomResourceDefinitions(), crds...); err != nil {
+			klog.Errorf("failed to bootstrap CRDs: %v", err)
+			return false, nil // keep retrying
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("failed to bootstrap CRDs into root workspace: %w", err)
+	}
+
+	// bootstrap non-crd resources
```
Comment (ncdc) [inline config/helpers/bootstrap.go:86]: What's this for?

**Review by ncdc — CHANGES_REQUESTED** [review]

File: config/helpers/bootstrap.go (hunk 2 above)
Comment (sttts) [reply config/helpers/bootstrap.go:83]: yes, we would

File: config/helpers/bootstrap.go (hunk 1 above)
Comment (sttts) [reply config/helpers/bootstrap.go:92]: I guess this is for line 95, ignoring the error

File: config/helpers/bootstrap.go (hunk 4 above)
Comment (sttts) [reply config/helpers/bootstrap.go:86]: E.g. workspace types, or any other kind of resources we need like roles/rolebindings.

Comment (ncdc) [reply config/helpers/bootstrap.go:86]: Ok, so mostly forward-looking at this point?

**Comment by sttts** [comment]
All adressed.

@ncdc CVO = OpenShift's cluster version operator

**Comment by ncdc** [comment]
LGTM. Will wait for @marun as well, since you've been looking at this also.

**Review by marun — APPROVED** [review]

File: config/helpers/bootstrap.go (hunk 4 above)
Comment (sttts) [reply config/helpers/bootstrap.go:86]: yes
